Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/BarcodeDetectorImplementation.mm

```diff
@@ -23,28 +23,205 @@
+    result.reserveInitialCapacity(formats.size());
+    for (auto format : formats)
+        result.add(format);
+    return result;
+}
+
+BarcodeDetectorImpl::BarcodeDetectorImpl(const BarcodeDetectorOptions& barcodeDetectorOptions)
+    : m_requestedBarcodeFormatSet(convertRequestedBarcodeFormatSet(barcodeDetectorOptions.formats))
 {
 }
 
 BarcodeDetectorImpl::~BarcodeDetectorImpl() = default;
 
+static RetainPtr<VNDetectBarcodesRequest> request()
+{
+    // It's important that both getSupportedFormats() and detect() use a VNDetectBarcodesRequest that is
+    // configured the same way. This function is intended to make sure both VNDetectBarcodesRequests are
+    // configured accordingly.
+
+    auto result = adoptNS([VNDetectBarcodesRequest new]);
+    configureRequestToUseCPUOrGPU(result.get());
+    return result;
+}
+
 void BarcodeDetectorImpl::getSupportedFormats(CompletionHandler<void(Vector<BarcodeFormat>&&)>&& completionHandler)
 {
-    completionHandler({ });
+    NSError *error = nil;
+    NSArray<VNBarcodeSymbology> *supportedSymbologies = [request() supportedSymbologiesAndReturnError:&error];
+
+    BarcodeFormatSet barcodeFormatsSet;
+    barcodeFormatsSet.reserveInitialCapacity(supportedSymbologies.count);
+    for (VNBarcodeSymbology symbology in supportedSymbologies)
+        barcodeFormatsSet.add(convertSymbology(symbology));
+
+    Vector<BarcodeFormat> barcodeFormatsVector;
+    barcodeFormatsVector.reserveInitialCapacity(barcodeFormatsSet.size());
+    for (auto barcodeFormat : barcodeFormatsSet)
+        barcodeFormatsVector.uncheckedAppend(barcodeFormat);
+
+    std::sort(std::begin(barcodeFormatsVector), std::end(barcodeFormatsVector));
+
+    completionHandler(WTFMove(barcodeFormatsVector));
 }
 
-void BarcodeDetectorImpl::detect(Ref<ImageBuffer>&&, CompletionHandler<void(Vector<DetectedBarcode>&&)>&& completionHandler)
+void BarcodeDetectorImpl::detect(Ref<ImageBuffer>&& imageBuffer, CompletionHandler<void(Vector<DetectedBarcode>&&)>&& completionHandler)
 {
-    completionHandler({ });
+    auto nativeImage = imageBuffer->copyNativeImage();
+    if (!nativeImage) {
+        completionHandler({ });
+        return;
+    }
+
+    auto platformImage = nativeImage->platformImage();
+    if (!platformImage) {
+        completionHandler({ });
+        return;
+    }
+
+    auto request = ShapeDetection::request();
+
+    if (m_requestedBarcodeFormatSet) {
+        NSMutableSet<VNBarcodeSymbology> *requestedSymbologies = [NSMutableSet setWithCapacity:m_requestedBarcodeFormatSet->size()];
+        for (auto barcodeFormat : *m_requestedBarcodeFormatSet) {
+            for (auto symbology : convertBarcodeFormat(barcodeFormat))
+                [requestedSymbologies addObject:symbology];
+        }
+        request.get().symbologies = requestedSymbologies.allObjects;
+    }
+
+    auto imageRequestHandler = adoptNS([[VNImageRequestHandler alloc] initWithCGImage:platformImage.get() options:@{ }]);
```

Comment:
why do we work with CGImages when the framework supports pixel buffers?
```
- (instancetype)initWithCVPixelBuffer:(CVPixelBufferRef)pixelBuffer options:(NSDictionary<VNImageOption, id> *)options;
- (instancetype)initWithCVPixelBuffer:(CVPixelBufferRef)pixelBuffer orientation:(CGImagePropertyOrientation)orientation options:(NSDictionary<VNImageOption, id> *)options;
-  (instancetype)initWithCVPixelBuffer:(CVPixelBufferRef)pixelBuffer
                            depthData:(AVDepthData*)depthData
                          orientation:(CGImagePropertyOrientation)orientation
                              options:(NSDictionary<VNImageOption, id> *)options;
```

The data comes from a camera I presume, so we could pass the CVPixelBuffer from the camera stream directly to the vision framework

Reply:
Yep! This is a good idea, but I think it's beyond the scope of this initial implementation.

This initial implementation keeps things simple by just having all the various formats drawn into an ImageBuffer, and then we do the recognition on the ImageBuffer. This could be improved; we could skip the copy and just run the recognition directly on the source, but that requires a lot more code, because there are 8 or so different forms the input could be in. So I think we should do that, but not in this patch.

Reply:
That sounds reasonable to me